DBSCAN++: Run DBSCAN on 100x larger datasets, up to 100x faster in subsampling

jenniferjang opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
See https://github.com/scikit-learn/scikit-learn/issues/17650.


#### What does this implement/fix? Explain your changes.
Instead of calculating the pairwise distances for all pairs of points to obtain nearest-neighbor graphs for DBSCAN, the `subsampled` option only calculates nearest-neighbor graphs for a fraction of points uniformly selected at random. The nearest-neighbor graphs are calculated only for subsampled points, but every point in the dataset may be a neighbor. This means that we will have a subset of core samples, but our research shows that clustering performance is similar.

With this implementation, we saw over 10x speedup on datasets where DBSCAN actually finished running. For datasets where DBSCAN did not finish running, subsampled DBSCAN is able to run on 100x larger datasets.

Performance is on a c5d.24xlarge Amazon EC2 instance with 96 vCPUs and 196 GB RAM. For all runs, `eps = 1` and `min_samples = 10`. Dataset was an equal mixture of four multivariate normals. We compare the results using the clustering metrics [adjusted RAND score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_rand_score.html) and [adjusted mutual information score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_mutual_info_score.html).
<img width="684" alt="image" src="https://github.com/user-attachments/assets/d91d6572-ad1f-473f-b387-635e875c760f" />

See: [Jang, J. and Jiang, H. "DBSCAN++: Towards fast and scalable density clustering". Proceedings of the 36th International Conference on Machine Learning, 2019.](https://proceedings.mlr.press/v97/jang19a.html)


#### Any other comments?
Per the November scikit-learn monthly meeting notes: 
<img width="735" alt="image" src="https://github.com/user-attachments/assets/1a0352be-7dc4-482e-ad5a-2e41f99ac860" />


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
DBSCAN++

Commit:
Updated in response to tests

Commit:
Updated docstring to pass test

kno10 commented:
Note that quality differences depend a lot on the data.
If you have synthetic data (think of make_blobs - or the "equal mixture of four multivariate normals" in your case), downsampling will be just fine. But this is not a realistic use case!
But the more complex the data gets and the less redundancy you have, the more downsampling can hurt.

I am not really sure about a use case where you (A) are fine with working on a subsample and only extending this to the remaining points, and (B) actually need to extend the labels to every object. Primarily, it is an explorative technique to help you understand the data, in such cases you can just work with a manageable subsample of the data right from the start.

jenniferjang commented:
> Note that quality differences depend a lot on the data. If you have synthetic data (think of make_blobs - or the "equal mixture of four multivariate normals" in your case), downsampling will be just fine. But this is not a realistic use case! But the more complex the data gets and the less redundancy you have, the more downsampling can hurt.
> 
> I am not really sure about a use case where you (A) are fine with working on a subsample and only extending this to the remaining points, and (B) actually need to extend the labels to every object. Primarily, it is an explorative technique to help you understand the data, in such cases you can just work with a manageable subsample of the data right from the start.

Hi Erich, thanks for your reply.

Addressing your comment:

“In such cases you can just work with a manageable subsample of the data right from the start.” I think there’s a slight misunderstanding of how our subsampling works. It’s not downsampling the entire dataset and then clustering the samples. It’s only subsampling the points that are core sample candidates, but still using the epsilon-neighborhood graph on the whole dataset to determine which points are core samples, and then clustering the entire dataset to the smaller number of core points. It only works because we use the whole dataset to determine the core points. 

As long as you choose samples from each cluster, you’re likely to be able to detect it – redundancy isn’t required. If we were to only run DBSCAN on a downsample, the examples in the minority clusters may not be correctly identified as core samples, and the cluster may not be detected.

We show in our paper that this works on 13 real (not synthetic) datasets, as well as being robust across hyperparameters, in many cases outperforming normal DBSCAN because it’s not as sensitive to border points.

From our paper:

![image](https://github.com/user-attachments/assets/0e1fac1e-5811-42c8-81c9-3bcb09a690c7)
Summary of open source datasets used. Includes dataset size (n), number of features (D), number of clusters (c), and the (m) used.

![image](https://github.com/user-attachments/assets/f08a738c-7e0d-447b-9d27-a9c3e2dce255)
Highest scores for each dataset and algorithm. The first row is the adjusted RAND index and the second row the adjusted mutual information. The standard error over 10 runs is given in parentheses for subsampled DBSCAN with uniform sampling. Each algorithm was tuned across a range of eps with `min_samples = 10`. We used subsample values of 0.1, 0.2, or 0.3. We see that DBSCAN + uniform sampling performs better than DBSCAN on 17 out of 22 metrics.

jenniferjang commented:
Hi @kno10, let me know what you think! Thanks!

kno10 commented:
None of these data sets (up to 60000 instances) is really a major scalability issue.
And the larger ones (MNIST) do not really cluster in a useful way with DBSCAN at all!

Can you prove any approximation guarantee for your method?

In fact, it is even a warning sign if your method outperforms exact DBSCAN by much more than a standard deviation - but here I'd argue that simply the labels do not reflect clusters, and hence the evaluation is meaningless.
And clustering has a *huge* evaluation problem, because we abuse classification labels all the time.

But again: I do not know any actual use case of having to cluster large data sets "exactly"; *standard* subsampling to, e.g., 1000 instances is usually enough for explorative data analysis. So your method IMHO targets a very small niche: what *applications* exist where standard subsampling is not enough, but that the exact result is not necessary and that fast approximate neighbor search (e.g., with HNSW) is not acceptable either.

(Which is not meant to say that users might now want this - often users do not want to subsample because they fear they might might lose something, and want to use clustering for classification... then only to be disappointed...)

jenniferjang commented:
> None of these data sets (up to 60000 instances) is really a major scalability issue. And the larger ones (MNIST) do not really cluster in a useful way with DBSCAN at all!
> 
> Can you prove any approximation guarantee for your method?
> 
> In fact, it is even a warning sign if your method outperforms exact DBSCAN by much more than a standard deviation - but here I'd argue that simply the labels do not reflect clusters, and hence the evaluation is meaningless. And clustering has a _huge_ evaluation problem, because we abuse classification labels all the time.
> 
> But again: I do not know any actual use case of having to cluster large data sets "exactly"; _standard_ subsampling to, e.g., 1000 instances is usually enough for explorative data analysis. So your method IMHO targets a very small niche: what _applications_ exist where standard subsampling is not enough, but that the exact result is not necessary and that fast approximate neighbor search (e.g., with HNSW) is not acceptable either.
> 
> (Which is not meant to say that users might now want this - often users do not want to subsample because they fear they might might lose something, and want to use clustering for classification... then only to be disappointed...)

In the paper, we do show that our modification of DBSCAN has similar theoretical guarantees as the original DBSCAN. At the last monthly scikit-learn meeting, we discussed this change and agreed that if it is simple, there should be room for it as many people were bumping the [original thread](https://github.com/scikit-learn/scikit-learn/issues/17650). Do you think we can get it in for their use cases and see if it works?

jenniferjang commented:
@kno10 friendly ping! Thanks!
